src/shortestpaths/johnson.jl

```diff
@@ -0,0 +1,96 @@
+
+"""
+    struct JohnsonState{T, U}
+An [`AbstractPathState`](@ref) designed for Johnson shortest-paths calculations.
+"""
+struct JohnsonState{T<:Real, U<:Integer} <: AbstractPathState
+    dists::Matrix{T}
+    parents::Matrix{U}
+end
+
+
+
+@doc_str """
+johnson_shortest_paths(g, distmx=weights(g); parallel = false)
+Use the [Johnson algorithm](https://en.wikipedia.org/wiki/Johnson%27s_algorithm)
+to compute the shortest paths between all pairs of vertices in graph `g` using an
+optional distance matrix `distmx`.
+If the parameter parallel is set true, dijkstra_shortest_paths will run in parallel.
+Parallel bellman_ford_shortest_paths is currently unavailable
+Return a [`LightGraphs.JohnsonState`](@ref) with relevant
+traversal information.
+Behaviour in case of negative cycle depends on bellman_ford_shortest_paths.
+Throws NegativeCycleError() if a negative cycle is present.
+### Performance
+Complexity: O(|V|*|E|)
+Prefered over Floyd Warshall when graph is sparse.
+If the graph is undirected (and has positive weights) then use Dijkstra.
+Undirected graph with negative weights will throw negative cycle error.
+### Dependencies from LightGraphs
+bellman_ford_shortest_paths
+parallel_multisource_dijkstra_shortest_paths
+parallel_multisource_dijkstra_shortest_paths
+"""
+function johnson_shortest_paths(
+    g::AbstractGraph{U},
+    distmx::AbstractMatrix{T} = weights(g);
+    parallel::Bool = false
+) where T<:Real where U<:Integer
+
+    nvg = nv(g)
+    #Change when parallel implementation of Bellman Ford available
+    wt_transform = bellman_ford_shortest_paths(g, vertices(g), distmx).dists
+    #Some matrices are immutable
+    new_distmx = convert(SparseMatrixCSC, distmx) 
```

Comment:
Are you sure this will work sanely with `DefaultDistance`? `weights(g::AbstractSimpleGraph)` simulates a fully-populated (dense) matrix. Converting it to sparse doesn't buy you anything and for large graphs you're going to run out of memory.

Reply:
I though SparseMatrix would make sense because Johnson is used for Sparase Graphs
Floyd Warshall is used for dense graphs.

An idea I had was create a single all_pairs_shortest_path program that works as follows:

If all weights are positive
 &ensp; Use Dijkstra
else
 &ensp; if graph is dense
 &ensp; &ensp; Use Floyd Warshall
  &ensp; else
 &ensp; &ensp;  Use Johnson

Hence, the default distance will be be covered using Dijkstra.